src/motion_model.py

```diff
@@ -1,225 +1,246 @@
+    # angle changed. Then update as normal.
+    def update_point(self, pt: tuple, index: int) -> tuple:
+        if random.randint(0, 5) == 4:
+            self.vel_angle[index] = random.uniform(0, 2 * pi)
+        return super().update_point(pt, index)
+
+
+class Viscek(BilliardMotion):
+
+    def __init__(self, dt, boundary: Boundary, n_int_sensors, sensing_radius):
+        super().__init__(dt=0, boundary=boundary, vel=1, n_int_sensors=n_int_sensors)
+        self.radius = sensing_radius
+        self.large_dt = dt
+
+    @staticmethod
+    def dist(pt1, pt2):
+        return norm(array(pt1) - array(pt2))
+
+    def eta(self):
+        return (pi / 12) * random.uniform(-1, 1)
+
+    def update_points(self, old_points: list, dt: float) -> list:
+        self.dt = dt
+        offset = len(self.boundary)
+        new_points = self.boundary.points \
+            + [self.update_point(pt, offset + n) for n, pt in enumerate(old_points[offset:])]
+
+        if self.dt != self.large_dt:
+            return new_points
+
+        for i, pti in enumerate(old_points[offset:]):
+            index_list = [j for j, ptj in enumerate(old_points[offset:]) if self.dist(pti, ptj) < self.radius
+                          and (self.radius < ptj[0] < 1 - self.radius)
+                          and (self.radius < ptj[1] < 1 - self.radius)]
+            # cluster_center = (np.mean([pti[j][0] for j in index_list]), np.mean([pti[j][1] for j in index_list]))
+            if index_list:
+                self.vel_angle[i+offset] = (mean([self.vel_angle[j+offset] for j in index_list]) + self.eta()) % (2 * pi)
+
+        return new_points
+
+
+
+## Implement D'Orsogna Collective Motion Model
+# Each sensor has a potential determined by it's distance to other sensors nearby which is used to compute dv/dt.
+# This motion model was originally developed to roughly emulate the dynamics of a large school of fish, for example.
+# Sensors are attracted to one another within a certain distance, but within a much shorter distance, they begin repel
+# one another.
+class Dorsogna(MotionModel):
+    def __init__(self, dt, boundary, max_vel, n_int_sensors, sensing_radius, eta_scale_factor, DO_coeff):
+        super().__init__(dt, boundary)
+        self.velocities = np.random.uniform(-max_vel, max_vel, (n_int_sensors, 2))
+        self.n_sensors = n_int_sensors
+        self.boundary = boundary
+        self.sensing_radius = sensing_radius
+        self.eta = eta_scale_factor * sensing_radius
+        if len(DO_coeff) != 4:
+            raise ValueError("Not enough parameters in DO_coeff")
+        self.DO_coeff = DO_coeff
+
+    def update_point(self, pt: tuple, index: int) -> tuple:
+        #Not used
+        return pt
+
+    ## This updates the velocities of each point before calling the reflect_point function for the given boundary
+    # If the point is outside the boundary, the reflect_point function for the boundary will reflect it accordingly.
```

Comment:
This function reflects points off of the domain elastically, and updates the velocity accordingly.